### PR TITLE
Use `iris.FUTURE.save_split_attrs = True` to remove iris warning

### LINKED ISSUE #2376: Lots of iris warnings in output since iris v3.8.0
Since iris v3.8.0, our output log is full of warnings like this:

```
2024-03-19 17:24:45,708 UTC [1445139] WARNING /work/bd0854/b309141/miniforge3/envs/esm/lib/python3.11/site-packages/iris/fileformats/netcdf/saver.py:2860: IrisDeprecation: Saving to netcdf with legacy-style attribute handling for backwards compatibility.
This mode is deprecated since Iris 3.8, and will eventually be removed.
Please consider enabling the new split-attributes handling mode, by setting 'iris.FUTURE.save_split_attrs = True'.
  warn_deprecated(message)
```

It would be nice to fix this in time for our v2.11.0.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

Closes #2376

Note: this is **not** fully backwards-compatible, as this changes the way iris writes attributes (some attributes are moved from global to local). I don't think this is a problem, but might be good to keep in mind.

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] ~~Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)~~ see above
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
I have no idea why the tests are failing. The test is complaining about a missing `tracking_id` attribute, but the [code](https://github.com/ESMValGroup/ESMValCore/blob/2cf5723b0c4652dfa9c1e68849fbb39831332852/tests/integration/recipe/test_recipe.py#L92) that is supposed to write this has never been called with a given `tracking_id`. Why did this not fail before? Did iris save tracking_ids automatically?

**Comment by valeriupredoi** [comment]
no, it's still there, but it's malformed:
```
valeriu@valeriu-PORTEGE-Z30-C:~/ESMValCore$ ncdump -h /tmp/pytest-of-valeriu/pytest-117/test_diagnostic_task_provenanc0/input/cmip5/output1/CCCma/CanESM2/historical/yr/ocnBgchem/Oyr/r1i1p1/*/chl_Oyr_CanESM2_historical_r1i1p1_2000_2009.nc
netcdf chl_Oyr_CanESM2_historical_r1i1p1_2000_2009 {
dimensions:
	dim0 = UNLIMITED ; // (0 currently)
variables:
	double unknown(dim0) ;
		unknown:tracking_id = 0LL ;

// global attributes:
		:Conventions = "CF-1.7" ;
}
```

as compared to without changes here:
```
valeriu@valeriu-PORTEGE-Z30-C:~/ESMValCore$ ncdump -h /tmp/pytest-of-valeriu/pytest-116/test_diagnostic_task_provenanc0/input/cmip5/output1/CCCMA/CanESM2/historical/yr/ocnBgchem/Oyr/r1i1p1/*/chl_Oyr_CanESM2_historical_r1i1p1_2000_2009.nc
netcdf chl_Oyr_CanESM2_historical_r1i1p1_2000_2009 {
dimensions:
	dim0 = UNLIMITED ; // (0 currently)
variables:
	double unknown(dim0) ;

// global attributes:
		:tracking_id = 0LL ;
		:Conventions = "CF-1.7" ;
}
```

File: esmvalcore/preprocessor/_io.py [hunk 1]
```diff
@@ -35,6 +35,8 @@
     'alternative_dataset',
 }
 
+iris.FUTURE.save_split_attrs = True
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_io.py:38]: luckily this is the only change in code (apart from environment changes/pins) so this is the one that's failing that test :grin:

**Review by valeriupredoi — CHANGES_REQUESTED** [review]

**Comment by valeriupredoi** [comment]
OK Manu, dug deeper in this (no pub for me this Friday): that tracking id is now (as in, using the call to iris' `save_split_attr`) an attribute of the netCDF4 variable:
```
iris attributes: {'Conventions': 'CF-1.7', 'tracking_id': 0}

netCDF4.Dataset variable items:
dict_items([('unknown', <class 'netCDF4._netCDF4.Variable'>
float64 unknown(dim0)
    tracking_id: 0
unlimited dimensions: dim0
current shape = (0,)
filling on, default _FillValue of 9.969209968386869e+36 used)])
```

-> it shouldn't be, it should be a global attribute of the file, not of the variable, as is the case when not using the FUTURE call:

```
iris attributes good cube: {'tracking_id': 0, 'Conventions': 'CF-1.7'}
<class 'netCDF4._netCDF4.Dataset'>
root group (NETCDF4 data model, file format HDF5):
    tracking_id: 0
    Conventions: CF-1.7
    dimensions(sizes): dim0(0)
    variables(dimensions): float64 unknown(dim0)
    groups: 
netCDF4 variable items good cube  dict_items([('unknown', <class 'netCDF4._netCDF4.Variable'>
float64 unknown(dim0)
unlimited dimensions: dim0
current shape = (0,)
filling on, default _FillValue of 9.969209968386869e+36 used)])
```

I think you should tell the iris folk the function is not doing the right thing :beer:

**Comment by bouweandela** [comment]
Attributes are now split between local and global, see https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.cube.html#iris.cube.CubeAttrsDict.__setitem__

**Comment by schlunma** [comment]
Yes, as Bouwe mentioned, this behavior of iris is expected. I was confused since it looked like the files are created without a `tracking_id`; however, the actual file that is tested is created in the [`patched_datafinder`](https://github.com/ESMValGroup/ESMValCore/blob/main/tests/integration/conftest.py#L112) fixture **with** a tracking id.

`iris.cube.Cube.attributes` will contain local and global attributes, so we are fine here. However, `netCDF4` does distinguish between the two. Thus, when reading the attributes for the provenance [here](https://github.com/ESMValGroup/ESMValCore/blob/6a984083255ef474dc7814f8621b4c7d922500ce/esmvalcore/_provenance.py#L209-L211), the `tracking_id` in the new code is skipped (since it has been written as a local attribute by the iris function). I fixed this by writing `tracking_id` as a global attribute for the test files in https://github.com/ESMValGroup/ESMValCore/pull/2398/commits/996ca364058ee681ec52f06d473242f01b9f2ba5.

**Comment by bouweandela** [comment]
@chrisbillowsMO It looks like we forgot to label this pull request with the backwards incompatible change. Could you please update the changelog for v2.11 so this is listed under backward incompatible changes?

**Comment by chrisbillowsMO** [comment]
> @chrisbillowsMO It looks like we forgot to label this pull request with the backwards incompatible change. Could you please update the changelog for v2.11 so this is listed under backward incompatible changes?

Will do!

**Pull request merged**